Bhima data collector

lomamech opened the pull request:
- Implente Client side for Data Collector
- Implete API for Data Collector Management
- Extend service color js adding new color
- Implement integration test for dataCollectorManagement
- Implement E2E test for dataCollectorManagement

This module makes it possible to configure the different types of the
choice lists, the lists of choices can be in the form of trees.
- Implement client side of choice list management
- Implement API for choice list Management
- Add integration and E2E for choice List Management

This module allows to add the elements of the form of a data collector,
the elements of a form have a type, some elements can have filter
compared to other elements
- Add new component for Data Collector
- Add new component for select Survey Form Type Select
- Add new component for Select Survey List Select
- Implement client side for Survey Form management
- Implement API for survey form manangement
- Add Integration and End to End test for Survey Form management

- Implementation of APIs related to the filling of forms as well as the
  display of survey data,

- Implementation of links between patient-related forms on the patient
  management page,

- Implementing the data collection report with a small search engine

- Implementation of components for Choicelist Multipke Select, Patient
  Medical Sheet, Survey Form Select,

- Adding Integration and End to End test for Displaying Metadata and
  Filling of survey Form

closes #3863

lomamech committed:
improvement('Service color')
- Sanitaze service color

lomamech committed:
Sanitaze code

lomamech committed:
improvement(Violations found)
- Sanitaze code and add comment eslint-disable no-eval and eslint-disable no-unused-vars

lomamech committed:
test(Fill Form)
- Fix incorrect property in E2E test for Fill Form

jniles commented on client/src/modules/survey_form/modals/survey_form.modals.html:
```diff
@@ -0,0 +1,206 @@
+      </div>
+    </div>
+
+    <div ng-if="SurveyFormModalCtrl.selectList" style="background-color:#ededed; padding-left: 10%; padding-right: 3%">
```
I'm not a fan of this padding - it looks a bit off.  There is a bootstrap class that is pretty close to what you want though - it is called `.well`.  See how it works below.

![8fW4Iki2dZ](https://user-images.githubusercontent.com/896472/64627532-b939f480-d3e7-11e9-8cf8-dc9c2c14c907.gif)

Would you be willing to change these `styles` to `class="well"`?

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+                    <span class="fa fa-file-pdf-o"></span> <span translate>DOWNLOADS.PDF</span>
+                  </a>
+                </li>
+                <li role="separator" class="divider"></li>    
```
Let's comment out the "separator" until the rest are done.  It makes the dropdown menu look a bit funny:

![DisplayMetadataExtraButton](https://user-images.githubusercontent.com/896472/64680718-3b70fa00-d476-11e9-91ff-026ab71a7a46.PNG)

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+              <ul uib-dropdown-menu role="menu" class="dropdown-menu-right">
+                <li role="menuitem">
+                  <a target='_blank'
+                    ng-href="/data_kit/report/?{{ DisplayMetadataCtrl.download(
```
Instead of having this large `DisplayMetadataCtrl.download()` call, it would be better to move this into the controller in a function called `downloadPDF()`.  This href will become `ng-href="/data_key/report?{{DisplayMetadataCtrl.downloadPDF()"`.

And then in your controller you can make a function:
```js
ctrl.downloadPDF = () => {
  return ctrl.download('pdf', ctrl.changes, ctrl.collectorId, /* etc ... */ );
};
```

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+        </button>
+      </div>
+
+      <div ng-if="DisplayMetadataCtrl.patientData" class="toolbar-item">
```
You have two `ng-if="DisplayMetadataCtrl.patientData"` next to each other.  Can you merge them?

jniles commented on client/src/modules/display_metadata/display_metadata.service.js:
```diff
@@ -0,0 +1,116 @@
+      renderer : type,
+      changes,
+      lang : Languages.key,
+      patient_uuid : patientUuid || null,
```
Why do you specify `patient` and `patientUuid` as two different parameters?

lomamech replied:
The parameter 'patient'  contains patient information for the purpose of no longer querying the database for its information

jniles replied:
Can you just use `patient.uuid`?

jniles commented on client/src/modules/display_metadata/display_metadata.service.js:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  function listSurveyformtype() {
+    const url = ''.concat('listSurveyformtype');
```
`const url = ''.concat('listSurveyformtype');` is the same as `const url = 'listSurvveyformType'`.  Go ahead and just use a plain string.

jniles commented on client/src/modules/display_metadata/display_metadata.service.js:
```diff
@@ -0,0 +1,116 @@
+  service.displayFilters = displayFilters;
+  service.download = download;
+
+  function download(type, changes, id, filterClient, patientUuid, patient) {
```
This function has a lot of parameters.  Can you just make it receive an `opts` object instead?  It would look something like this?

```js
function download(type, opts) {
  const options = Object.assign({}, opts,  { renderer : type });
  return $httpParamSerializer(options);
}
```

This would be easier for other developers to work with since they will not have to remember the order of each parameter.  It will also be more flexible if you need to change something in the future.

jniles commented on client/src/modules/display_metadata/display_metadata.service.js:
```diff
@@ -0,0 +1,116 @@
+      const multipleChoiceLength = Object.keys(search.multipleChoice).length;
+      if (multipleChoiceLength) {
+        Object.keys(search.multipleChoice).forEach((key) => {
+          survey.forEach(item => {
```
You are using `survey.forEach()` a lot in this function!  Let's make lookups more efficient by making a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) from the `survey` array.

Basically:
```js
const surveyMap = new Map(survey.map(item => ([item.name, item]));
```

Now, instead of writing:
```js
survey.forEach(item => {
  if (item.name === key) {
     // do something 
  }
});
```
you can write:
```js
const item = surveyMap.get(key);
```

If the item doesn't exist, `surveyMap.get()` will return `undefined`.

jniles commented on client/src/modules/display_metadata/display_metadata.service.js:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  function displayFilters(survey, search) {
+    let filters = ``;
```
I'm not a big fan of using `+=` to add to a string.  Could you instead use an array, do `array.push()` for each string, and then later to `array.join('')` to turn it into a larger string?

My logic is that it lets you use array functions (`forEach`, `map`, etc) on each item if you need them later.  It also looks a bit cleaner.

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+      </div>
+
+      <div class="toolbar-item">
+        <button
```
We have a component for this - [bhFilterToggle](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/js/components/bhFilterToggle.js).  You could use it here to reduce code:

```html
 <bh-filter-toggle on-toggle="DisplayMetadataCtrl.toggleFilter()">
 </bh-filter-toggle>
```

lomamech replied:
I could not use the bh-filter-toggle component just because in normal use the elements that use this component come from the elements defined in advance in the FilterService service, whereas the elements of the structure of the grid comes from the database

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+<div class="flex-util" style="overflow: auto" ng-if="DisplayMetadataCtrl.collector">
+  <span data="filter">
+    <span class="label" style="text-shadow: {{ DisplayMetadataCtrl.collector.color }} 5px 3px 8px; font-size: 15px; color: black">
+      <span> {{ DisplayMetadataCtrl.collector.label }} </span>
```
Do you need these extra `<span>`s?

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+  </span>
+
+  <span ng-if="DisplayMetadataCtrl.filterElements" data="filter">
+    <span class="label" ng-class="{ 'label-primary' : true, 'label-emphasize' : true }">
```
Do you need an `ngClass` like this `ng-class="{ 'label-primary' : true, 'label-emphasize' : true }"`?   Why not just make these regular HTML classes?
```html
<span class="label label-primary label-emphasize">
```

jniles commented on client/src/modules/display_metadata/display_metadata.html:
```diff
@@ -0,0 +1,121 @@
+
+  <span ng-if="DisplayMetadataCtrl.patientData" data="filter">
+    <span class="label"  style="font-size: 15px; color: black; text-shadow: gray 4px 1px 10px;">
+      <span>:{{ DisplayMetadataCtrl.patient.display_name }} / {{ DisplayMetadataCtrl.patient.reference}} </span>
```
It looks like you have extra spans here as well.

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+    enableSorting     : true,
+    showColumnFooter  : true,
+    onRegisterApi     : onRegisterApiFn,
+    enableHorizontalScrollbar : uiGridConstants.scrollbars.WHEN_NEEDED,
```
This is cool 👍

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+    fastWatch         : true,
+    flatEntityAccess  : true,
+    enableSorting     : true,
+    showColumnFooter  : true,
```
Do you do anything with the column footer?  I don't see it being used in my tests.

lomamech replied:
Thanks, I add the number of rows 
![image](https://user-images.githubusercontent.com/8035588/64964678-6f438980-d893-11e9-9188-c935d12ca031.png)

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+
+  if ($state.params.id && $state.params.patient) {
+    // Prevent a non-patient form from being used to collect non-patient data
+    vm.patientData = true;
```
Since this is a boolean, let's call it `hasPatientData`.

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+  vm.remove = remove;
+
+  function edit(data) {
+    if (!$state.params.patient) {
```
We already checked for this in the startup code with `hasPatientData`.  Could we use that variable instead of `$state`?

```js
if (hasPatientData) {
  // do x
} else {
  // do y
}
```

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+    if (cache.collector && !vm.collectorId && !$state.params.patient) {
+      vm.collectorId = cache.collector.id;
+    }
+    const changesLength = Object.keys(vm.changes).length;
```
Let's make this into a boolean:
```js
const hasNoChanges = Object.keys(vm.changes).length === 0;
```
Then we can use it in a nice check:
```
if (cache.changes && hasNoChanges) {
  // do x
}
```

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+        .then((survey) => {
+          vm.filterElements = DisplayMetadata.displayFilters(survey, vm.changes);
+
+          vm.params = {
```
Does `vm.params` depend on the result of `SurveyForm.read()`?  If not, let's move it out:
```js
vm.params = {
  data_collector_management_id : vm.collectorId,
  changes : vm.changes,
};

SurveyForm.read(/* etc ...*/)
```

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+          vm.gridOptions.columnDefs = data.columns;
+          vm.gridOptions.data = data.surveyData;
+
+          return DataCollectorManagement.read(vm.collectorId);
```
You have a nice promise chain here that sends a request for:
 1. `SurveyForm.read()`
 2. `DisplayMetadata.read()`
 3. `DataCollectorManagement.read()`

However, it doesn't look like any of these actually depend on one another.  So, let's send them in parallel using `Promise.all()`!

```js
vm.params = {/* etc  */};
return Promise.all([
  SurveyForm.read(null, { data_collector_management_id : vm.collectorId }),
  DisplayMetadata.read(null, vm.params),
  DataCollectorManagement.read(vm.collectorId),
])
.then(([survey, data, collector]) => {
  // process all queries...
});
```

lomamech replied:
@jniles, i try this options but the displaying became too slow

jniles commented on client/src/modules/display_metadata/display_metadata.js:
```diff
@@ -0,0 +1,208 @@
+        changes : vm.changes,
+      };
+
+      DisplayMetadata.read(null, vm.params)
```
See if you can do a similar `Promise.all()` system here.

lomamech replied:
![image](https://user-images.githubusercontent.com/8035588/64969611-26dc9980-d89c-11e9-9f10-a13b2b830506.png)
The displaying became very very too slow

jniles reviewed (commented):
@lomamech,

This PR is incredible - it represents a lot of work, and it performs very well.  Well done!

As discussed, I've limited my review to the `DisplayMetaData` modules for the moment, and mostly the graphical user interface.

Let me know if you have any questions!

lomamech committed:
imporvement(Survey Form)
- Disabled the button add when survey form is not selected

lomamech committed:
improvement(Fill Form)
-Add the number of submissions for each survey form

lomamech committed:
refactor(Download PDF)
- Improve display metadata, reduice the function downloadD
- avoid apostrophes in the search parameter

lomamech committed:
refactor(Data Collector)
- Added the listHint property to display information about the component
  selection lists bhChoiceListMultipleSelect and bhChoiceListSelect

- Using function downloadPDF and Using an object instead of a function with
  multiple parameters

- Using Mao from the survey array for mere efficient

lomamech committed:
Improve and sanitaze code

lomamech committed:
Resolve conflict s3

jniles commented on client/src/js/components/bhChoiceListMultipleSelect.js:
```diff
@@ -0,0 +1,56 @@
+      listLabel        : '@?',
+      isTitle          : '<?',
+      isGroup          : '<?',
+      parentId         : '<',
```
It looks like the `parentId` is optional (line 36).  This should be `<?`.

jniles commented on client/src/js/components/bhChoiceListSelect.js:
```diff
@@ -0,0 +1,64 @@
+      listLabel        : '@?',
+      isTitle          : '<?',
+      isGroup          : '<?',
+      parentId         : '<',
```
This should also be optional `<?`.

lomamech committed:
improvement(Bhima data Collector)
- Clear and Sanitaze Code
closes #3863

jniles commented on client/src/modules/choices_list_management/modals/choices_list_management.modals.js:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  function closeModal() {
+    $state.transitionTo('choices_list_management');
```
Most of our `$state` changes use `$state.go()`.  Can you use that there?

jniles commented on client/src/modules/choices_list_management/templates/action.cell.tmpl.html:
```diff
@@ -0,0 +1,37 @@
+
+    <ul data-row-menu="{{row.entity.label}}" class="dropdown-menu-right" bh-dropdown-menu-auto-dropup uib-dropdown-menu>
+      <li class="dropdown-header text-ellipsis" style="max-width:250px;">
+        <a> {{row.entity.label}} </a>
```
Why does this need to be wrapped in a `<a>` tag?  Can you just put it in plain `<span>` tag instead?

jniles commented on client/src/modules/data_collector_management/modals/data_collector_management.modals.js:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  function closeModal() {
+    $state.transitionTo('data_collector_management');
```
`$state.go()` should be sued here as well.

jniles commented on client/src/modules/choices_list_management/templates/gridName.tmpl.html:
```diff
@@ -0,0 +1,8 @@
+  <span ng-if="row.entity.is_title" style="text-transform: uppercase">
+    {{ row.entity.label}}
+  </span>
+  <span ng-if="!row.entity.is_title" style="font-weight:normal; font-style: italic; text-align: right">
```
Do you need the `font-style:italic` if you have wrapped the element in `<i>`?

jniles commented on client/src/modules/data_collector_management/modals/data_collector_management.modals.html:
```diff
@@ -0,0 +1,98 @@
+    <!-- description -->
+    <div class="form-group" ng-class="{ 'has-error' : DataCollectorManagementForm.$submitted && DataCollectorManagementForm.description.$invalid }">
+      <label class="control-label" translate>FORM.LABELS.DETAILS</label>
+      <div>
```
Why do you need to wrap this text area in a `<div>`?

jniles commented on client/src/modules/data_collector_management/modals/data_collector_management.modals.html:
```diff
@@ -0,0 +1,98 @@
+    <!-- version_number -->
+    <div class="form-group" ng-class="{ 'has-error' : DataCollectorManagementForm.$submitted && DataCollectorManagementForm.version_number.$invalid }">
+      <label class="control-label" translate>FORM.LABELS.VERSION</label>
+      <div>
```
It looks like you are wrapping a lot of elements in `<div>`s - is there a reason for this?

jniles commented on client/src/modules/data_collector_management/modals/data_collector_management.modals.js:
```diff
@@ -0,0 +1,74 @@
+
+function DataCollectorManagementModalController($state, DataCollectorManagement, Notify, AppCache, Color) {
+  const vm = this;
+  const cache = AppCache('AccountReferenceModal');
```
This shouldn't be `AccountReferenceModal`.

jniles commented on client/src/modules/survey_form/modals/survey_form.modals.js:
```diff
@@ -0,0 +1,108 @@
+
+function SurveyFormModalController($state, SurveyForm, Notify, AppCache, DataCollectorManagement) {
+  const vm = this;
+  const cache = AppCache('AccountReferenceModal');
```
This probably shouldn't be `AccountReferenceModal`.

jniles commented on client/src/modules/survey_form/modals/survey_form.modals.js:
```diff
@@ -0,0 +1,108 @@
+  function submit(surveyForm) {
+    vm.hasNoChange = surveyForm.$submitted && surveyForm.$pristine && !vm.isCreating;
+    if (surveyForm.$invalid) { return null; }
+    if (!surveyForm.$dirty) { return null; }
```
Instead of `!surveyForm.$dirty`, go ahead and use `surveyForm.$pristine`.  It is the same thing, but without a `!`.

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.js:
```diff
@@ -0,0 +1,230 @@
+  function closeModal() {
+    const transitionTo = vm.updateMode ? 'display_metadata' : 'fill_form';
+    if (!vm.stateParams.patient) {
+      $state.transitionTo(transitionTo);
```
Can you use `$state.go()` here?

jniles commented on client/src/modules/patients/record/patient-record.html:
```diff
@@ -132,6 +132,15 @@
+      <div class="row">
+        <!-- Medical Sheet -->
+        <div class="col-xs-12">
+          <bh-patient-medical-sheet patient-uuid="PatientRecordCtrl.patient.uuid">
```
I propose we name this something else - "medical sheet" makes it sound like it has something to do with the patient's medical information, but it is actually just data collections about the patient.

Let's call this `<bh-patient-data-collection-forms>` or something like that.

lomamech replied:
the use of the term "medical sheet" simply wanted to mention that this area is dedicated to medical information of patients.

Logically, the forms that should appear on this zone should only be those collecting medical information

jniles reviewed (commented):
@lomamech Great work on making these changes.

I'm still working through most of the code, but I've made a few more comments.  Take a look :)

:+1:

lomamech committed:
refactor(Bhima Data Collector)
- Update component bhChoiceListSelect bhChoiceListMultipleSelect change
  params parentId in '<' to '<?'
- Using $state.go instead $state.transitionTo
- Using <span> instead not necessary tag <a> ... </a>
- Using $pristine instead !<--->.$dirty
- Sanitaze and Clear Code

lomamech committed:
refactor(Bhima Data Collector)
- Improved form selection component, added disable option when viewing
  patient data
- Reformated the research params by the using of components bh-filters
- Improved data display module, removal of entangled requests in the
  controller
-

lomamech committed:
Clear and Sanitaze Code

jniles commented on client/src/modules/choices_list_management/choices_list_management.service.js:
```diff
@@ -0,0 +1,100 @@
+function ChoicesListManagementService(Api) {
+  const service = new Api('/choices_list_management/');
+
+  service.formatStore = formatStore;
```
Hmm... looks like we have identical code in the [AccountService](https://github.com/IMA-WorldHealth/bhima/blob/master/client/src/modules/accounts/accounts.service.js#L115).  Is there any way we can move this into a shared service so that we don't have to duplicate code?

jniles commented on client/src/modules/data_collector_management/modals/data_collector_management.modals.html:
```diff
@@ -0,0 +1,94 @@
+          ng-true-value="1"
+          ng-false-value="0"
+          ng-model="DataCollectorManagementModalCtrl.dataCollector.is_related_patient">
+        <span translate>
```
This `translate` tag needs to be on the `<strong>` not the `<span>`.

jniles commented on client/src/js/components/bhPatientMedicalSheet.js:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    function toggleLoadingIndicator() {
+      $ctrl.loading = false;
```
Toggle loading indicator always sets "loading" to be "false".  Shouldn't it turn loading on and off?

jniles commented on client/src/js/components/bhPatientMedicalSheet.js:
```diff
@@ -0,0 +1,108 @@
+    $ctrl.gridApi = {};
+    $ctrl.filterEnabled = false;
+    $ctrl.toggleFilter = toggleFilter;
+    $ctrl.consult = consult;
```
Just FYI - for functions, you don't need to put them in the `$onInit` call.  The `$onInit()` is most for variables that are passed in and/or need to be initialized with each component.

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+    // S'il sont multiples afin de les affecter dans des tableaux
+    form.forEach(f => {
+      data.forEach(d => {
+        if ((f.id === d.survey_form_id) && (f.type === '4')) {
```
What is `type === '4'`?  Why not `type === '7'`?

lomamech replied:
![image](https://user-images.githubusercontent.com/8035588/66905516-e89edb00-effd-11e9-99f7-e39e30325761.png)
It's form multiple value, look in bhima.sql in table survey_form_type

jniles replied:
Can you instead use the label?  It will make it much easier for a developer to read.  Like this:
```js
f.label === 'FORM.LABELS.SELECT_MULTIPLE' // now all future developers knows what this means!
```

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+      });
+    });
+
+    form.forEach(f => {
```
You are doing two `.forEach()` loops here that you could combine.    Just move lines 26 through 28 to the start of the next loop and it should save you a whole loop.

lomamech replied:
The reason is that in the first loop, we initialize a table for all the elements of the type 4 form, and in the second loop we simply push the elements in this array

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+    // S'il sont multiples afin de les affecter dans des tableaux
+    form.forEach(f => {
+      data.forEach(d => {
+        if ((f.id === d.survey_form_id) && (f.type === '4')) {
```
It looks like you use this check more than once.  Can you make a function that tells us what it is?  Like this:
```js
function isValidSurveyForm(form, data) {
  return form.id === data.survey_form_id && form.type === '4';
}
```

Then you can just do:
```js
form.forEach(f => {
  data.forEach(d => {
    if (isValidSurveyForm(f, d)) {
      // do something.
    }
  });
});

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+          dataSurveyForm[f.name].push(parseInt(d.value, 10));
+        } else if ((f.id === d.survey_form_id) && (f.type !== '4') && (f.type !== '7')) {
+          // Single choice list management
+          d.value = (f.type === '1') ? parseFloat(d.value) : d.value;
```
Aha, so `f.type` gives the question type.  Could you set those as constants in this service or in the `bhConstants` constant?

In this way, it will be much more readable.  For example:
```js
d.value = (f.type === QUESTION_DATE_TYPE) ? new Date(d.value) : d.value;
d.value = (f.type === QUESTION_CHOICE_LIST_TYPE) ? coolfunction() : null;
```

lomamech replied:
@jniles, All type come from table survey_form_type
![image](https://user-images.githubusercontent.com/8035588/66907766-c3ac6700-f001-11e9-9014-1318f26df386.png)

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+          d.value = (f.type === '3') ? parseInt(d.value, 10) : d.value;
+
+          // format date
+          d.value = (f.type === '6') ? new Date(d.value) : d.value;
```
You are writing something like `d.value = (f.type === x) ? onething() : d.value` over and over again.  Let's  extract this code into a separate parser function that uses a [`switch`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch) statement.  Something like this:

```js
d.value = parseDataValue(d.value, f.type);


// somewhere 
function parseDataValue(value, type) {
  switch(type) {
    case '1':
      return parseInt(value, 10);
    case '4':
      return new Date(value);
    default:
      return value;
  }
}
```

If you make this a function _and_ you implement the constants suggestion above, you will get a really nice looking function:
```js
function parseDataValue(value, type) {
  switch(type) {
    case QUESTION_INTEGER_TYPE:
      return parseInt(value, 10);
    case QUESTION_DATE_TYPE:
      return new Date(value);
    // etc...
    default:
      return value;
  }
}
```
My syntax for the switch function may be wrong, be sure to test it out.

jniles commented on client/src/modules/fill_form/fill_form.service.js:
```diff
@@ -0,0 +1,86 @@
+    * is incorrectly evaluates the return function a null value
+  */
+
+  function formatConstraint(constraint) {
```
I like this function - putting code in a function like this makes it possible to write good tests for it.  Well done! 👍 

Did you write a couple unit tests for this function? 😄

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+    </ol>
+  </div>
+
+  <div style="background-color: {{ FillFormModalCtrl.dataCollector.color }}; border-radius : 25px ; text-align: center; text-decoration-color: #ffffff; text-shadow: rgb(255, 255, 255) 5px 3px 8px;">
```
Can you use [`ngStyle`](https://docs.angularjs.org/api/ng/directive/ngStyle) here instead of hacking in the style?

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+    <h4><strong ng-if="FillFormModalCtrl.patient.display_name"> {{ FillFormModalCtrl.patient.display_name }} / {{ FillFormModalCtrl.patient.reference}} ({{ FillFormModalCtrl.patient.sex }}) </strong></h4>
+  </div>
+
+  <div class="modal-body" popover-append-to-body="true" style="overflow-y: scroll; min-height:300px; max-height:600px; ">    
```
Where did this "popover-append-to-body" come from?  What does it do?

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+
+      <div ng-if="!formItem.constraint || $eval(formItem.constraintCheck)">
+        <!-- Type text -->
+        <div ng-if="formItem.typeForm === 'text'" class="form-group" ng-class="{ 'has-error' : FillFormForm.$submitted && FillFormForm.{{ formItem.name }}.$invalid }">
```
As discussed together, I think that you can simply do `FillFormForm[formItem.name].$invalid` and save yourself a bit of parsing.

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+        <div ng-if="formItem.typeForm === 'text'" class="form-group" ng-class="{ 'has-error' : FillFormForm.$submitted && FillFormForm.{{ formItem.name }}.$invalid }">
+          <label class="control-label"> {{ formItem.label }} </label>
+          <span style="color: #cecece"><em> {{ formItem.hint }} </em></span>
+          <div>
```
Do you need this `<div>` here?

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+            <span ng-if="formItem.required">
+              <input name="{{ formItem.name }}" ng-model="FillFormModalCtrl.form[formItem.name]" autocomplete="off" ng-maxlength="90" class="form-control" required>
+            </span>
+            <span ng-if="!formItem.required">
```
Why not use [`ngRequired`](https://docs.angularjs.org/api/ng/directive/ngRequired)?

jniles commented on client/src/modules/fill_form/modals/fill_form.modals.html:
```diff
@@ -0,0 +1,250 @@
+        </div>
+
+        <!-- TextArea -->
+        <div ng-if="formItem.typeForm === 'text_area'">
```
Since you are doing a lot of `formItem.typeForm === "x"`, could you use an [`ngSwitch`](https://docs.angularjs.org/api/ng/directive/ngSwitch)?  It would probably make the code easier to read.

jniles commented on client/src/modules/reports/generate/dataKit/dataKit.config.js:
```diff
@@ -0,0 +1,133 @@
+          if (vm.reportDetails.multipleChoice[key].length) {
+            for (let i = 0; i < vm.reportDetails.multipleChoice[key].length; i++) {
+              vm.choicesLists.forEach(list => {
+                if (list.id === vm.reportDetails.multipleChoice[key][i]) {
```
So .. it looks like you are taking a list of ids and turning them into a list of `label`s.  So, you are trying to do:
```js
var ids = [3, 3, 5, 7, 1]
// do something
console.log(ids) // => ['formA', 'formA', 'formZ', 'formX', 'formT']
```

Let's try to clean this up a bit.  Here is what I think the basic structure could look like:
```js
angular.forEach(vm.reportDetails.mulitpleChoice, (values, key) => {
  // values is an array.
  values.forEach(value => {
    vm.choicesLists.forEach(list => {
      if (list.id === value) {
         // replace value with list.label
      }
    });
  });
});
```

Let's see if we can go one step further:
```js
// create a way to map id -> label for the choice list
let choiceListMap = vm.choicesList.reduce((map, item) => {
  map[item.id] = item.label;
}, {});

angular.forEach(vm.reportDetails.mulitpleChoice, (values, key) => {
  // unfortunately, angular.map() doesn't exist for object.  We have to use an ugly forEach
  vm.multipleChoice[key] = values.map(value => choiceListMap[value]);
});
```

I think that should do it.

First, we create an object to make choiceList ids map to their labels.  This takes up memory but removes the `choiceLists.forEach()` loop.

Then, we use `angular.forEach()` to loop through the object.  This removes all the `Object.keys()` calls and checks for `.length`.  

Try it out and see if I've missed anything.

jniles reviewed (commented):
@lomamech, a few more comments.  Your changes from last time look good.

lomamech committed:
Resolve Conflict with branch master

lomamech committed:
Rebase with branch Master

lomamech committed:
Rebase with branch master

lomamech committed:
Rebase with branch master

lomamech committed:
Rebase with branch master

lomamech committed:
Rebase with branch Master

lomamech committed:
Rebase with branch master

lomamech committed:
improvement(Data Collector)
- Update placeHolder message in component bhSurveyTypeSelect
- Using function parseDataValue and use structure swith for manage type of
  data
- Using ng-required instead to duplicate block of code
- Update survey form modals, hide the selection filter for Select Multiple
- Use the basic structure for manage vm.reportDetails.multipleChoice value
  in form of search in Report Data kit

lomamech committed:
Sanitaze Code

lomamech committed:
improvement(Data Collector)
- Added the option to include the patient selection component in the data
  collection form

- Display of the name of the user as well as that of the patient if the
  form is associated with the patient

lomamech committed:
test(Validation configuration form survey)
- Implementation of a service to validate the variable name parameter when
  configuring forms,
- Implementation of the unit test for the validation of the variable
  parameter name
- Implementation of the E2E test for the validation of the variable
  name parameter when configuring formulators

lomamech committed:
Sanitaze and clear code

lomamech committed:
Sanitaze and Clear Code

lomamech committed:
Rebase continue with branch Master

lomamech committed:
Resolve conflict in Migration file

lomamech committed:
Comment migrate.sql

lomamech committed:
Resolve conflict
